platform/src/main/python/dlpx/virtualization/platform/migration_helper.py

```diff
@@ -0,0 +1,327 @@
+        return std_string
+
+    @staticmethod
+    def __validate_migration_id(migration_id, impl_name):
+        # First validate that the id is a string
+        if not isinstance(migration_id, basestring):
+            raise MigrationIdIncorrectTypeError(migration_id, impl_name)
+
+        # Next check if the id is the right format
+        if not PlatformUpgradeMigrations.MIGRATION_ID_REGEX.match(
+                migration_id):
+            raise MigrationIdIncorrectFormatError.from_fields(
+                migration_id, impl_name,
+                PlatformUpgradeMigrations.MIGRATION_ID_REGEX.pattern)
+
+    @staticmethod
+    def __standardize_migration_id_to_array(migration_id, impl_name):
+        # Split on the period and convert to integer
+        array = [int(i) for i in migration_id.split('.')]
+
+        #
+        # We cannot allow a migration id of essentially '0' because otherwise
+        # there would be no way to add a migration that goes before this.
+        #
+        if not any(array):
+            raise MigrationIdIncorrectFormatError(
+                "The migration id '{}' used in the function '{}' cannot be"
+                " used because a 0 migration id is not allowed.".format(
+                    migration_id, impl_name))
+
+        # Next we want to trim all trailing zeros so ex: 5.3.0.0 == 5.3
+        while array:
+            if not array[-1]:
+                # Remove the last element which is a zero from array
+                array.pop()
+            else:
+                break
+
+        return array
+
+    def get_sorted_ids(self):
+        # First sort the migration ids
+        self.__migration_ids.sort()
+
+        # Then convert all these arrays to the usual string format.
+        return [
+            '.'.join(str(i) for i in migration_id)
+            for migration_id in self.__migration_ids
+        ]
+
+    def get_repository_impls_to_exec(self, migration_id_list):
+        return self.__get_impls(migration_id_list, self.get_repository_dict())
+
+    def get_source_config_impls_to_exec(self, migration_id_list):
+        return self.__get_impls(migration_id_list,
+                                self.get_source_config_dict())
+
+    def get_linked_source_impls_to_exec(self, migration_id_list):
+        return self.__get_impls(migration_id_list,
+                                self.get_linked_source_dict())
+
+    def get_virtual_source_impls_to_exec(self, migration_id_list):
+        return self.__get_impls(migration_id_list,
+                                self.get_virtual_source_dict())
+
+    def get_snapshot_impls_to_exec(self, migration_id_list):
+        return self.__get_impls(migration_id_list, self.get_snapshot_dict())
+
+    @staticmethod
+    def __get_impls(migration_id_list, impl_dict):
+        return_list = []
+        for migration_id in migration_id_list:
+            # Should only add the function if the id exists in the map.
```

Comment:
Should it be an error to ask for an ID that isn't in the list?  I can't think of a use case where it would be okay/desired to do this, but I'm not confident I've thought of all cases.

Reply:
hmm actually.. this should not error out because the migration id list includes all migration ids (in the set) where as the impl dict will only include ones that were written specifically for that object type (say repo or source config). Therefore we do just want to skip if the id doesn't exist in the dict.

Reply:
Okay, I think I get it. `migtation_id_list` represents "all migrations to run, regardless of type", and "impl_dict" represents "all migrations for this type, regardless of whether they are going to be run or not".  Is that right?

Is there a check someplace for what happens when the engine asks to run a migration ID that does not appear in ANY of the datatype-specific lists?  But it might be that we don't need to worry about that?

Reply:
so we could definitely do that but currently we don't. I don't think it's important enough to do right now either. I can file a follow on bug that talks about this and put it into a future phase.

Reply:
https://github.com/delphix/virtualization-sdk/issues/117